### PR TITLE
Add HWM settings in Ledger

### PR DESCRIPTION
**Pull request by ajinkyaraj-23**
- [x] Add HWM to N_data persistent storage
- [x] Add HWM to settings in nano home layout
- [x] Add HWM switch to stax settings page
- [x] Remove migrated_to_tenderbake

### CODE REVIEW

File: test/snapshots/stax/test_review_home/BIP32_ED25519_tz1VKyZ3RFDwTkrz5LKcTc6fcYqZj6pvsyA7/app_context.png
Comment (spalmer25) [inline test/snapshots/stax/test_review_home/BIP32_ED25519_tz1VKyZ3RFDwTkrz5LKcTc6fcYqZj6pvsyA7/app_context.png:1]: Why does the exit button change ?
Does the test touch-exit function still work ?

File: test/snapshots/stax/test_review_home/BIP32_ED25519_tz1VKyZ3RFDwTkrz5LKcTc6fcYqZj6pvsyA7/description.png
Comment (spalmer25) [inline test/snapshots/stax/test_review_home/BIP32_ED25519_tz1VKyZ3RFDwTkrz5LKcTc6fcYqZj6pvsyA7/description.png:1]: Please make a separate commit for the Developer/Copyright change

File: test/test_instructions.py [hunk 1]
```diff
@@ -149,6 +175,13 @@ def both():
         tezos_navigator.settings.next()
         backend.wait_for_screen_change()
         screen("description")
+        tezos_navigator.settings.next()
+        backend.wait_for_screen_change()
+        screen("hwm_status_on")
```
Comment (spalmer25) [inline test/test_instructions.py:180]: I do not see any screen named `hwm_status_on`

File: src/ui_bagl.c [hunk 2]
```diff
@@ -74,10 +79,35 @@ UX_FLOW(ux_idle_flow,
         &ux_version_step,
         &ux_chain_id_step,
         &ux_authorized_key_step,
+        &ux_settings_step,
         &ux_hwm_step,
```
Comment (spalmer25) [inline src/ui_bagl.c:83]: I think the settings should be placed right before the exit screen.

File: src/ui_nbgl.c [hunk 3]
```diff
@@ -81,41 +92,77 @@ static bool navigation_cb_baking(uint8_t page, nbgl_pageContent_t* content) {
...
+static void controls_callback(int token, uint8_t index) {
+    UNUSED(index);
+    bool switch_value;
+    if (token == HWM_ENABLED_TOKEN) {
+        // HWM switch touched
+        // toggle the switch value
+        switch_value = !N_data.hwm_disabled;
+        // store the new setting value in NVM
+        nvm_write((void*) &N_data.hwm_disabled, &switch_value, 1);
+    }
```
Comment (spalmer25) [inline src/ui_nbgl.c:136]: Prefer to use `globals.toggle_hwm` which uses `UPDATE_NVRAM`.

**Review by spalmer25 — COMMENTED** [review]
The `hwm` settings does not have any impact on the app, is that on purpose?

**Comment by ajinkyaraj-23** [comment]
> The `hwm` settings does not have any impact on the app, is that on purpose?

The app behaviour changes are in progress. I just got the display and button changes done.

File: src/ui_bagl.c (hunk 2 above)
Comment (ajinkyaraj-23) [reply src/ui_bagl.c:83]: Sure. Makes sense.

File: src/ui_nbgl.c (hunk 3 above)
Comment (ajinkyaraj-23) [reply src/ui_nbgl.c:136]: Sure.

File: test/snapshots/stax/test_review_home/BIP32_ED25519_tz1VKyZ3RFDwTkrz5LKcTc6fcYqZj6pvsyA7/app_context.png
Comment (ajinkyaraj-23) [reply test/snapshots/stax/test_review_home/BIP32_ED25519_tz1VKyZ3RFDwTkrz5LKcTc6fcYqZj6pvsyA7/app_context.png:1]: Here i chaged the base setting "SETTINGS_BUTTON_ENABLED" which causes the exit button layout to change into back button layout.

Comment (spalmer25) [reply test/snapshots/stax/test_review_home/BIP32_ED25519_tz1VKyZ3RFDwTkrz5LKcTc6fcYqZj6pvsyA7/app_context.png:1]: I also have these settings buttons changes on other PR (See [this job](https://github.com/trilitech/ledger-app-tezos-baking/actions/runs/8541455677/job/23427634398?pr=92))
It seems this is a `Stax` version update, here is the PR that fix it : #93

File: test/utils/navigator.py [hunk 4]
```diff
@@ -81,13 +82,15 @@ class TezosNavigator(metaclass=MetaScreen):
 
     use_case_home       = UseCaseHome
     use_case_settings   = UseCaseSettings
+    use_case_sub_settings = UseCaseSubSettings
```
Comment (spalmer25) [inline test/utils/navigator.py:85]: You can use the already defined `settings` instead of reintroducing a new one.

File: src/ui_nbgl.c [hunk 5]
```diff
@@ -81,41 +92,72 @@ static bool navigation_cb_baking(uint8_t page, nbgl_pageContent_t* content) {
...
+        case 2:
+            switches[HWM_ENABLED_TOKEN_ID].initState = (nbgl_state_t) (!hwm_disabled);
+            switches[HWM_ENABLED_TOKEN_ID].text = "High Watermark";
+            switches[HWM_ENABLED_TOKEN_ID].subText = "Track high watermark\n in Ledger";
+            switches[HWM_ENABLED_TOKEN_ID].token = HWM_ENABLED_TOKEN;
+            switches[HWM_ENABLED_TOKEN_ID].tuneId = TUNE_TAP_CASUAL;
+            content->type = SWITCHES_LIST;
+            content->switchesList.nbSwitches = SETTINGS_SWITCHES_NB;
+            content->switchesList.switches = (nbgl_layoutSwitch_t*) switches;
+            break;
```
Comment (spalmer25) [inline src/ui_nbgl.c:117]: Personal opinion:
I think that the Copyright/version should remain at the end and that the parameters should be more accessible.
`Baking data > Settings > Copyright`

File: test/test_instructions.py [hunk 6]
```diff
@@ -129,31 +133,61 @@ def both():
...
         backend.wait_for_screen_change()
         screen("app_context")
-        tezos_navigator.settings.next()
+        tezos_navigator.sub_settings.next()
+        backend.wait_for_screen_change()
+        screen("description")
+        tezos_navigator.sub_settings.next()
+        backend.wait_for_screen_change()
+        screen("hwm_status")
+        # TODO: switch hwm status off and check the screen.
```
Comment (spalmer25) [inline test/test_instructions.py:183]: I think we should check the `hwm` toggle here (we can probably use `Layouts.ChoiceList`)

File: test/utils/client.py [hunk 7]
```diff
@@ -80,11 +80,11 @@ class Hwm:
     """Class representing app high water mark."""
 
     highest_level: int
-    highest_round: Optional[int]
+    highest_round: int
 
     def __init__(self,
                  highest_level: int,
-                 highest_round: Optional[int] = None,
+                 highest_round: int = 0,
```
Comment (spalmer25) [inline test/utils/client.py:87]: Let's not have a default value for the round, because it's important to be explicit when we create a `HWM`.

File: test/utils/navigator.py [hunk 8]
```diff
@@ -242,10 +242,8 @@ def check_app_context(self,
             self.sub_settings.next()
             self.backend.wait_for_screen_change()
             self.assert_screen("hwm_status", snap_path=snap_path)
-            self.settings.multi_page_exit()
-            self.backend.wait_for_screen_change()
-            self.assert_screen("home_screen", snap_path=snap_path)
-
+            self.sub_settings.exit()
+            self.backend.wait_for_home_screen()
```
Comment (spalmer25) [inline test/utils/navigator.py:246]: I'm not sure why such a change has been added here.

**Review by spalmer25 — COMMENTED** [review]
I'm glad we've finally got rid of the dependency on what came before tenderbake

File: src/ui_bagl.c [hunk 9]
```diff
@@ -46,11 +46,15 @@ typedef struct {
     char chain_id[CHAIN_ID_BASE58_STRING_SIZE];
     char authorized_key[PKH_STRING_SIZE];
     char hwm[MAX_INT_DIGITS + 1u + MAX_INT_DIGITS + 1u];
+    char hwm_status[HWM_STATUS_SIZE];
```
[the remark this answers is not in the document]
Comment (ajinkyaraj-23) [reply src/ui_bagl.c:49]: How about we make it part of another issue. That will be more clear. Created a new[ issue](https://github.com/trilitech/ledger-app-tezos-baking/issues/94)

File: src/ui_bagl.c [hunk 10]
```diff
@@ -75,9 +80,34 @@ UX_FLOW(ux_idle_flow,
...
+
+// FLOW for the about submenu:
+// #1 screen: app info
+// #2 screen: back button to main menu
+UX_FLOW(ux_settings_flow, &ux_hwm_info, &ux_menu_back_step, FLOW_LOOP);
+
+void ui_settings(void) {
+    memcpy(home_context.hwm_status,
+           N_data.hwm_disabled ? "Disabled" : "Enabled",
+           sizeof(home_context.hwm_status));
```
[the remark this answers is not in the document]
Comment (ajinkyaraj-23) [reply src/ui_bagl.c:102]: Done.

File: test/utils/navigator.py (hunk 4 above)
Comment (ajinkyaraj-23) [reply test/utils/navigator.py:85]: Previous settings and sub_settings had different layout. So exit page and next previous was different. But now it seems they have same layout after the 3.19.0 upgrade.

Comment (ajinkyaraj-23) [reply test/utils/navigator.py:85]: See following in ragger 
```
    "UseCaseSettings": {
        "single_page_exit": BUTTON_LOWER_MIDDLE,
        "multi_page_exit": BUTTON_LOWER_LEFT,
        "previous": BUTTON_LOWER_MIDDLE,
        "next": BUTTON_LOWER_RIGHT,
    },
    "UseCaseSubSettings": {
        "exit": BUTTON_UPPER_LEFT,
        "previous": BUTTON_LOWER_LEFT,
        "next": BUTTON_LOWER_RIGHT,
    },
```

Comment (spalmer25) [reply test/utils/navigator.py:85]: Oh yes, you're right !
Now that we have the right version, we can use the `UseCaseSettings`:
https://github.com/LedgerHQ/ragger/blob/master/src/ragger/firmware/stax/positions.py#L206-L216

File: src/ui_nbgl.c (hunk 5 above)
Comment (ajinkyaraj-23) [reply src/ui_nbgl.c:117]: Done.

File: test/utils/client.py (hunk 7 above)
Comment (ajinkyaraj-23) [reply test/utils/client.py:87]: The round value is appended to response from the app for command get_hwm, if the protocol is upgraded to tenderbake. To remove default value 0 we will have to change behaviour of get_hwm in the app. What do you think? Should i not append round value to HWM if round value is zero. I feel keeping the round value seems more appropriate, we can indicate that HWM referes to block and round.

File: src/ui_bagl.c [hunk 11]
```diff
@@ -106,6 +136,10 @@ static bool calculate_baking_idle_screens_data(void) {
 
     TZ_ASSERT(hwm_to_string(home_context.hwm, sizeof(home_context.hwm), &N_data.hwm.main) >= 0,
               EXC_WRONG_LENGTH);
+    TZ_ASSERT(hwm_status_to_string(home_context.hwm_status,
+                                   sizeof(home_context.hwm_status),
+                                   &N_data.hwm_disabled) >= 0,
+              EXC_WRONG_LENGTH);
```
Comment (spalmer25) [inline src/ui_bagl.c:142]: Why do we need to update the `hwm_status` here?

File: test/utils/client.py [hunk 12]
```diff
@@ -195,9 +195,13 @@ class TezosClient:
 
     backend: BackendInterface
 
+
     def __init__(self, backend) -> None:
         self.backend = backend
 
+    def finger_touch(self, x: int = 0, y: int = 0, delay: float = 0.5) -> None:
+        self.backend.finger_touch(x,y,delay)
+
```
Comment (spalmer25) [inline test/utils/client.py:202]: This code is unused

File: src/ui_bagl.c (hunk 11 above)
Comment (ajinkyaraj-23) [reply src/ui_bagl.c]: We dont. removed it. Also rebased on main.

File: test/utils/client.py (hunk 12 above)
Comment (spalmer25) [reply test/utils/client.py:202]: Oh, that because the ChoiceList() needs a `BackendInterface` and not a `TezosClient`:
In https://github.com/trilitech/ledger-app-tezos-baking/blob/b169879f73b2575682a7debb7fe8eb67f0c1a9dd/test/utils/navigator.py#L117
The line should be:
```python
        self.layout_choice = ChoiceList(backend, firmware)
```
Sorry to have missed this one

Comment (ajinkyaraj-23) [reply test/utils/client.py]: Got it. Changed.

**Review by spalmer25 — APPROVED** [review]
LGTM
